enhance contracts query

Omarabdul3ziz opened the pull request:
### Description
separate contracts bills in a new query to speed up the contracts query

### Changes
- add two new endpoints `/contracts/contract_id` and `/contracts/contract_id/bills`
- add helper method for creating response with count pagination headers
- remove billing from contract response and remove the join with contract_bills table

### Related Issues
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/347
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/231

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

Omarabdul3ziz committed:
enhance contracts query:
- add two new endpoints `/contracts/contract_id` and `/contracts/contract_id/bills`
- add helper method for creatingResponse with count pagination headers
- remove billing from contract response and remove the join with contract_bills table
-

Omarabdul3ziz committed:
add swagger docs

mariobassem commented on grid-proxy/pkg/types/types.go:
```diff
@@ -259,13 +259,12 @@ type RentContractDetails struct {
-	Type       string            `json:"type"`
-	Details    interface{}       `json:"details"`
-	Billing    []ContractBilling `json:"billing"`
+	ContractID uint        `json:"contractId"`
```
can we use snake case names for json tags?

mariobassem commented on grid-proxy/internal/explorer/server.go:
```diff
@@ -467,6 +439,61 @@ func (a *App) getNodeGpus(r *http.Request) (interface{}, mw.Response) {
+// @Failure 500 {object} string
+// @Router /contracts/{contract_id} [get]
+func (a *App) getContract(r *http.Request) (interface{}, mw.Response) {
+	contractID := mux.Vars(r)["contract_id"]
```
what if contract_id is a invalid number, shouldn't we return  a bad request errror?
if so, i believe this contract id conversion should happen here, no?

Omarabdul3ziz replied:
yes indeed. the conversions happen first thing in the next called method `a.getContractData(contractID)` just tended to do the same as in `getNode` function :D

mariobassem commented on grid-proxy/internal/explorer/helpers.go:
```diff
@@ -24,6 +25,20 @@ func errorReply(err error) mw.Response {
 	}
 }
 
+func createResponse(totalCount uint, limit types.Limit) mw.Response {
```
we needed this indeed

mariobassem reviewed (changes requested):
great work!!
can you also add some tests for the billing reports endpoint?

Omarabdul3ziz committed:
add tests for new contract and bills endpoints

Omarabdul3ziz committed:
fix bills tests

mariobassem commented on grid-proxy/internal/explorer/db/postgres.go:
```diff
@@ -817,6 +816,47 @@ func (d *PostgresDatabase) GetContracts(filter types.ContractFilter, limit types
+		Select("amount_billed, discount_received, timestamp").
+		Where("contract_id = ?", contractID)
+
+	q = q.Limit(int(limit.Size)).
```
should we add some ordering for the retrieved bill reports?

mariobassem commented on grid-proxy/internal/explorer/helpers.go:
```diff
@@ -305,3 +320,46 @@ func (a *App) getNodeData(nodeIDStr string) (types.NodeWithNestedCapacity, error
+
+	bills := []types.ContractBilling{}
+	for _, report := range info {
+		bills = append(bills, contractBillFromDBContractBill(report))
```
since both types are basically the same type, i believe you don't need the conversion, no?

mariobassem commented on grid-proxy/pkg/types/types.go:
```diff
@@ -258,14 +258,53 @@ type RentContractDetails struct {
+	State      string          `json:"state"`
+	CreatedAt  uint            `json:"created_at"`
+	Type       string          `json:"type"`
+	Details    ContractDetails `json:"details"`
```
i believe that the details is creating some unneeded hassle, maybe we should change its type??

Omarabdul3ziz committed:
enhance and simplify decoding contract details

Omarabdul3ziz committed:
order the reports desc

Omarabdul3ziz committed:
resolve comments:
- add order by timestamp for reports on mock client
- remove unnecessary conversion

Omarabdul3ziz commented:
Note: re-requested some reviews just to test it will trigger the tests again. and it worked :+1:
